### DISCUSSION TITLE
Passport mechanism is over engineering and useless

### DISCUSSION THREAD

**Discussion by tugrul** [symfony/symfony#43276] [Q&A]
Previous authenticator approach was pretty simple and efficient. Passport mechanism looks pretty good fashion but what problem is solving with this design change?

Authenticator needs at least an UserInterface object and validate credentials. I think there would be an interface of that two methods on top of passport mechanism and it can be easily avoidable by who don't prefer to use this passport mechanism.

My actual authenticator implementation is like below and I feel passport mechanism is tangled in my feet.

```php
<?php

namespace App\Security;

use App\Entity\User;
use App\Form\UserLoginType;
use Doctrine\ORM\EntityManagerInterface;
use Symfony\Component\Form\FormFactoryInterface;
use Symfony\Component\HttpFoundation\RedirectResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
use Symfony\Component\Security\Core\Exception\CustomUserMessageAuthenticationException;
use Symfony\Component\Security\Core\Encoder\UserPasswordEncoderInterface;
use Symfony\Component\Security\Core\Security;
use Symfony\Component\Security\Core\User\UserInterface;
use Symfony\Component\Security\Core\User\UserProviderInterface;
use Symfony\Component\Security\Guard\Authenticator\AbstractFormLoginAuthenticator;
use Symfony\Component\Security\Http\Util\TargetPathTrait;

class EmailAuthenticator extends AbstractFormLoginAuthenticator
{
    use TargetPathTrait;

    private $entityManager;
    private $urlGenerator;
    private $passwordEncoder;
    private $loginForm;

    public function __construct(EntityManagerInterface $entityManager, UrlGeneratorInterface $urlGenerator,
                                UserPasswordEncoderInterface $passwordEncoder,
                                FormFactoryInterface $formFactory)
    {
        $this->entityManager = $entityManager;
        $this->urlGenerator = $urlGenerator;
        $this->passwordEncoder = $passwordEncoder;

        $this->loginForm = $formFactory->create(UserLoginType::class, []);
    }

    public function supports(Request $request)
    {
        if ('user-login' !== $request->attributes->get('_route')) {
            return false;
        }

        $this->loginForm->handleRequest($request);

        return $this->loginForm->isSubmitted() && $this->loginForm->isValid();
    }

    public function getCredentials(Request $request)
    {
        $credentials = $this->loginForm->getData();

        $request->getSession()->set(
            Security::LAST_USERNAME,
            $credentials['email']
        );

        return $credentials;
    }

    public function getUser($credentials, UserProviderInterface $userProvider)
    {
        $user = $this->entityManager->getRepository(User::class)
            ->findOneBy(['email' => $credentials['email']]);

        if (!$user) {
            throw new CustomUserMessageAuthenticationException('User does not exists');
        }

        return $user;
    }

    public function checkCredentials($credentials, UserInterface $user)
    {
        return $this->passwordEncoder->isPasswordValid($user, $credentials['password']);
    }

    public function onAuthenticationSuccess(Request $request, TokenInterface $token, $providerKey)
    {
        if ($targetPath = $this->getTargetPath($request->getSession(), $providerKey)) {
            return new RedirectResponse($targetPath);
        }

        return new RedirectResponse($this->urlGenerator->generate('homepage'));
    }

    protected function getLoginUrl()
    {
        return $this->urlGenerator->generate('user-login');
    }

    public function getLoginForm()
    {
        return $this->loginForm;
    }
}

```

**Comment by lyrixx** [comment]
> Previous authenticator approach was pretty simple and efficient. Passport mechanism looks pretty good fashion but what problem is solving with this design change?

simplification of the previous system for both symfony and your code

With the new system, your current authenticator will be even simpler.

**Comment by javiereguiluz** [comment]
Please also note that the new security system is lacking lots of docs. We're working on a complete rewrite of Symfony Security Docs (see https://github.com/symfony/symfony-docs/pull/15503). Once this is completed, we hope things like the benefits of the new system will be easier to understand.

**Reply by tugrul** [reply]
I'm able to understand idea during navigate source code. There were many checkpoint methods previous authentication approach. Now it has single checkpoint and this is good idea. But it force me to use Passport and its Pass collection. Maybe Passport can supply bunch of good solutions but this is not main goal of authentication flow. Main goal is fetch user & check credentials. How is making this things are not goal of authentication. 

I would see Passport is opt-in feature and would see `public function authenticate(Request $request): UserAuthenticateInterface` instead of `public function authenticate(Request $request): PassportInterface`

So `PassportInterface` can extend `UserAuthenticateInterface` and can provide itself as opt-in feature.

Other users like me, can prefer to implement `UserAuthenticateInterface` and can avoid to use Passport if they prefer.

**Reply by lyrixx** [reply]
So you prefer to check the password yourself instead of wrapping your user in Password?

so, IIRC, you say the "Passport mechanism is over engineering and useless" ​because you need to need to change one line of code by another one?

BTW, during authentication, some user want to do more than checking the password. For example: check the CSRF token, add rate limiting, check the user state (disabled) etc.

The new system is more flexible, and is closer to the reality (the word passport is chosen on purpose)

**Reply by chess-server** [reply]
@javiereguiluz: The documentation should include a section which shows at least one more complicated example how to convert from an Guard Authentication to the new strategy. I am adding information to the session (new columns in the session table!) and I would not like to loose these information when changing the system. From what I have seen so far, this would be the case. 

Please note I am not against improvements, but I am an amateur programmer with a large project (two instances of a chess server) and limited time. In the past I was struggling a lot with unclear documentation of symfony components. Once I did understand the philosophy, I could introduce many improvements into the code. An example: Last week(!), I found an interesting way to add tournament types to the different instances of the server using the EventDispatcher system. Until then I hardcoded the types into both instances. 

I would like to steepen the learning curve for the new security system (from which I hope I can find measures to prevent unwanted registrations from sites east of Europe).

**Comment by wouterj** [comment]
Passports do only one thing: **centralization of logic** .... and that's a big deal imho!

One of the major advantages of using a framework like Symfony (or any other open source framework) is that you can use code maintained and build by hundreds of developers. If one developer in India discovers a weak vulnerability in the code, it will be fixed for everyone around the world. With hundreds of developers, this means you'll get a lot safer code.
But.. you can only make code in the framework safer. So our goal is to move as much security critical code towards the framework, to secure all users from security vulnerabilities.

The code you shared in this question shows that every Symfony user was responsible for checking the password in their custom authenticator. This means that checking the password can no longer be improved by the community.
This might not look like much, but checking a password has resulted in quite some CVE releases in the past like [handling empty passwords](https://symfony.com/blog/cve-2017-11365-empty-passwords-validation-issue), but also [handling too long passwords](https://symfony.com/blog/cve-2013-5750-security-issue-in-fosuserbundle-login-form) (something I would never ever think about if Symfony hadn't fixed this for me). In fact, the password checking of the passport system uses [50 lines of code](https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Security/Http/EventListener/CheckCredentialsListener.php#L53-L100) to do what you did in a single line of code, because of checking all sorts of edge-cases.

On top of this, centralizing logic also makes it much easier to use some features. There is no need to figure out how to use the CSRF component, just add the `CsrfTokenBadge` to the passport with the csrf token and the passport system does the rest for you. Same with other features like login throttling. But also with third party features like MFA.

---

Finally, the new `AuthenticatorInterface` allowed to rewrite all internal authenticators (like form_login, json_login, remember_me, etc) to use this interface (something that was not possible with the old guard authenticator interface). This means that Symfony now uses the same interface internally that you would use to write a custom authenticator in your application. I hope this will activate some users to contribute their great custom authenticators to Symfony, meaning the security system will move forward (whereas the current security system didn't change much in 10 years).

**Reply by wouterj** [reply]
Btw, I'm not a big fan of people judging something to be "over engineering and useless" while they don't actually understand the reasoning behind it. Please, first do your research, then judge.

_edit: To clarify: the question is great; that's the research part (which is also why I tried to give it a lengty response), but I don't like the tone of the question title, which is judging_

**Comment by tugrul** [comment]
Modularity is good for some scenarios but also is bad some scenarios. Is it worth to make modular if there are small set of possible changes? I'm sharing two samples to see analogy. 

## Sample 1
I prefer to make in-place changes like this
```javascript
const answerText = 'yuck';
const shouldHaveUppercaseChar = Math.random() > 0.5;

const rgValidText = /^[a-z]{2}([a-zA-Z])[a-z]$/

function getValue(value) {
    const match = rgValidText.exec(value);

    if (match === null) {
        throw Error('invalid badge');
    }

    if (shouldHaveUppercaseChar) {
        const chars = value.split('');
        chars[2] = chars[2].toUpperCase();
        return chars.join('');
    }

    return value;
}

console.log(getValue(answerText));
```

## Sample 2
I'm able to add event listener but is it worth for this bulky code?
```javascript
const EventEmitter = require('events');

class Badge {
    value = null;

    constructor(value) {
        this.value = value;
    }

    getValue() {
        return this.value;
    }

    setValue(value) {
        this.value = value;
    }

    isValid() {
        return true;
    }
}

class LowercaseCharacterBadge extends Badge {

    isValid() {
        const charCode = this.value.charCodeAt(0);
        return charCode >= 97 && charCode <= 122;
    }

}

class MaybeUppercaseCharacterBadge extends Badge {

    valid = false;

    setValid(valid) {
        this.valid = valid;
    }

    isValid() {
        return this.valid;
    }

}

class Answer extends EventEmitter {
    badges = []

    constructor(...badges) {
        super();
        this.badges = badges;
    }

    getValue() {
        return this.badges.map(badge => {

            this.emit('badge', badge);

            if (!badge.isValid()) {
                throw Error('invalid badge');
            }

            return badge.getValue();

        }).join('');
    }
}

const answerText = 'yuck';
const shouldHaveUppercaseChar = Math.random() > 0.5;

const answer = new Answer(
    new LowercaseCharacterBadge(answerText[0]),
    new LowercaseCharacterBadge(answerText[1]),
    new MaybeUppercaseCharacterBadge(answerText[2]),
    new LowercaseCharacterBadge(answerText[3])
);

answer.on('badge', (badge) => {
    if (badge instanceof MaybeUppercaseCharacterBadge) {
        let value = badge.getValue();
        if (shouldHaveUppercaseChar) {
            value = value.toUpperCase();
            const charCode = value.charCodeAt(0);
            badge.setValid(charCode >= 65 && charCode <= 90);
            badge.setValue(value);
        } else {
            const charCode = value.charCodeAt(0);
            badge.setValid(charCode >= 97 && charCode <= 122);
        }
    }
});

console.log(answer.getValue());
```

**Reply by wouterj** [reply]
I'm not sure what you try to say here... Software design is context sensitive, meaning that the same design for one problem can be over engineering and for the other can be under engineering - that's the whole game of the field that we all work in.

I just gave you the 3 or 4 main reasons for us to believe the passport concept isn't over engineering. It would have been cool to have a discussion about these reasons, but I think it's good simply to agree to disagree if we compare authentication to lettercase checking of a string.

_Btw, your 2nd sample is also actively increasing the complexity of the badge system. There no reason to drop the regex to check the whole string in favor of char code checks per character and not using a `TransformToUpper(value, index)` badge to upper the 3rd character similarly to your 1st sample._
